test/test_transformers.py

```diff
@@ -2967,6 +2971,189 @@ def rand_nt(sequence_list, num_heads, head_dim):
+            torch._dynamo.reset()
+        torch.manual_seed(123)
+        bsz, num_heads, seq_len_q, seq_len_kv, head_dim = shapes
+        if causal_variant == CausalVariant.LOWER_RIGHT and seq_len_q > seq_len_kv:
+            self.skipTest(
+                "Lower right causal mask will produce NaNs in the output when seq_len_q > seq_len_kv!"
+            )
+
+        make_q_tensor = partial(make_tensor, SdpaShape(bsz, num_heads, seq_len_q, head_dim))
+        make_kv_tensor = partial(make_tensor, SdpaShape(bsz, num_heads, seq_len_kv, head_dim))
+
+        query, key, value = make_q_tensor(), make_kv_tensor(), make_kv_tensor()
+        query_prototype, key_prototype, value_prototype = query_key_value_clones(query, key, value)
+        attn_bias = CausalBias(causal_variant, seq_len_q, seq_len_kv)
+
+        pytorch_output = scaled_dot_product_attention(
+            query, key, value, attn_mask=attn_bias.materialize(device), dropout_p=0.0, is_causal=False
+        )
+        sdpa_op = (
+            torch.compile(scaled_dot_product_attention, fullgraph=True)
+            if compile
+            else scaled_dot_product_attention
+        )
+        sdpa_output = sdpa_op(
+            query_prototype,
+            key_prototype,
+            value_prototype,
+            attn_mask=attn_bias,
+            scale=None,
+            is_causal=False,
+            dropout_p=0.0,
+        )
+
+        dOut = torch.randn_like(pytorch_output)
+        pytorch_output.backward(dOut)
+        sdpa_output.backward(dOut)
+        atol, rtol = 1e-3, 1e-3
+        grad_atol, grad_rtol = 5e-3, 5e-3
+        torch.testing.assert_close(pytorch_output, sdpa_output, atol=atol, rtol=rtol)
+        torch.testing.assert_close(query.grad, query_prototype.grad, atol=grad_atol, rtol=grad_rtol)
+        torch.testing.assert_close(key.grad, key_prototype.grad, atol=grad_atol, rtol=grad_rtol)
+        torch.testing.assert_close(value.grad, value_prototype.grad, atol=grad_atol, rtol=grad_rtol)
+
+
+    @parametrize("shape", [SdpaShape(16, 16, 128, 16), SdpaShape(16, 16, 52, 32)])
+    @parametrize("compile", [True, False])
+    def test_tensor_bias(self, device, shape: SdpaShape, compile: bool):
+        if compile:
+            self.skipTest("Compiling torch_function_ not working")
+            torch._dynamo.reset()
+
+        make_tensor = partial(
+            torch.rand, shape, device=device, dtype=torch.float16, requires_grad=True
+        )
+        query, key, value = make_tensor(), make_tensor(), make_tensor()
+        query_prototype, key_prototype, value_prototype = query_key_value_clones(query, key, value)
+        attn_bias = TensorBias(
+            torch.rand(shape.batch, shape.num_heads, shape.seq_len, shape.seq_len, dtype=torch.float16, device=device)
+        )
+
+        pytorch_output = scaled_dot_product_attention(
+            query, key, value, attn_mask=attn_bias.materialize(device), dropout_p=0.0, is_causal=False
```

Comment:
This manual `materialize()` call is unnecessary right? I'd expect that the user never has to interact with materialization as a concept (unless they're implementing a custom attn bias) and that it would be tucked behind the abstraction

Reply:
Yeah exactly, doing this here just to enforce we are testing against OG sdpa